miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+        cmd("rustup", ["toolchain", "uninstall", "miri"]).dir(&self.miri_dir).run()?;
+        let args = [
+            "-n",
+            "miri",
+            "-c",
+            "cargo",
+            "-c",
+            "rust-src",
+            "-c",
+            "llvm-tools",
+            "-c",
+            "rustfmt",
+            "-c",
+            "clippy",
+        ]
+        .into_iter()
+        .chain(flags.iter().map(String::as_str))
+        .chain(["--", &new_commit]);
+
+        cmd("rustup-toolchain-install-master", args).dir(&self.miri_dir).run()?;
+        cmd("rustup", ["override", "set", "miri"]).dir(&self.miri_dir).run()?;
+        // Cleanup.
+        cmd("cargo", ["clean"]).dir(&self.miri_dir).run()?;
+        // Call `cargo metadata` on the sources in case that changes the lockfile
+        // (which fails under some setups when it is done from inside vscode).
+        let sysroot = cmd("rustc", ["--print", "sysroot"]).dir(&self.miri_dir).read()?;
+        let sysroot = sysroot.trim();
+        let manifest_path =
+            format!("{sysroot}/lib/rustlib/rustc-src/rust/compiler/rustc/Cargo.toml");
+        let _ =
+            cmd("cargo", ["metadata", "--format-version", "1", "--manifest-path", &manifest_path])
+                .dir(&self.miri_dir)
+                .read()?;
+        Ok(())
+    }
+
+    fn rustc_pull(&self, commit: Option<String>) -> Result<()> {
+        let commit = commit.map(Result::Ok).unwrap_or_else(|| {
+            let rust_repo_head =
+                cmd("git", ["ls-remote", "https://github.com/rust-lang/rust/", "HEAD"])
+                    .dir(&self.miri_dir)
+                    .read()?;
+            Ok(rust_repo_head
+                .split_whitespace()
+                .next()
+                .expect("Could not obtain Rust repo HEAD from remote.")
+                .trim()
+                .to_owned())
+        })?;
+        // Update rust-version file. As a separate commit, since making it part of
+        // the merge has confused the heck out of josh in the past.
+        std::fs::write(Path::new(&format!("{}/rust-version", self.miri_dir.display())), &commit)?;
+        const PREPARING_COMMIT_MESSAGE: &str = "Preparing for merge from rustc";
+        cmd("git", ["commit", "rust-version", "-m", PREPARING_COMMIT_MESSAGE])
+            .run()
+            .expect("FAILED to commit rust-version file, something went wrong");
+        // Fetch given rustc commit and note down which one that was
+        let fetch_url =
+            format!("http://localhost:8000/rust-lang/rust.git@{commit}{JOSH_FILTER}.git");
+        cmd("git", ["fetch", &fetch_url])
+            .run()
+            .expect("FAILED to fetch new commits, something went wrong");
+        const MERGE_COMMIT_MESSAGE: &str = "Merge from rustc";
+        cmd("git", ["merge", "FETCH_HEAD", "--no-ff", "-m", MERGE_COMMIT_MESSAGE])
+            .run()
+            .expect("FAILED to merge new commits, something went wrong");
```

Comment:
This is using `expect` a lot for things that can totally actually happen, e.g. a merge conflict. Please use proper error reporting instead of panics. Probably this should be using anyhow and then we can write
```suggestion
            .context("FAILED to merge new commits, something went wrong")?;
```